src/integrity-checker.js

```diff
@@ -196,52 +253,67 @@ export default class InstallationIntegrityChecker {
     return null;
   }
 
-  async _compareIntegrityFiles(
+  _compareIntegrityFiles(
     actual: IntegrityFile,
     expected: ?IntegrityFile,
     checkFiles: boolean,
-    locationFolder: string,
     workspaceLayout: ?WorkspaceLayout,
-  ): Promise<'OK' | IntegrityError> {
+  ): 'OK' | IntegrityError {
     if (!expected) {
       return 'EXPECTED_IS_NOT_A_JSON';
     }
+
     if (!compareSortedArrays(actual.linkedModules, expected.linkedModules)) {
       return 'LINKED_MODULES_DONT_MATCH';
     }
-    if (!compareSortedArrays(actual.flags, expected.flags)) {
+
+    let relevantExpectedFlags = expected.flags.slice();
+
+    // If we run "yarn" after "yarn --check-files", we shouldn't fail the less strict validation
+    if (actual.flags.indexOf('checkFiles') === -1) {
+      relevantExpectedFlags = relevantExpectedFlags.filter(flag => flag !== 'checkFiles');
+    }
+
+    if (!compareSortedArrays(actual.flags, relevantExpectedFlags)) {
       return 'FLAGS_DONT_MATCH';
     }
-    const actualPatterns = actual.topLevelPatterns.filter(p => {
-      return !workspaceLayout || !workspaceLayout.getManifestByPattern(p);
-    });
-    if (!compareSortedArrays(actualPatterns, expected.topLevelPatterns || [])) {
+
+    if (!compareSortedArrays(actual.topLevelPatterns, expected.topLevelPatterns || [])) {
       return 'PATTERNS_DONT_MATCH';
     }
+
     for (const key of Object.keys(actual.lockfileEntries)) {
       if (actual.lockfileEntries[key] !== expected.lockfileEntries[key]) {
         return 'LOCKFILE_DONT_MATCH';
       }
     }
+
     for (const key of Object.keys(expected.lockfileEntries)) {
       if (actual.lockfileEntries[key] !== expected.lockfileEntries[key]) {
         return 'LOCKFILE_DONT_MATCH';
       }
     }
+
     if (checkFiles) {
-      if (expected.files.length === 0) {
-        // edge case handling - --check-fies is passed but .yarn-integrity does not contain any files
-        // check and fail if there are file in node_modules after all.
-        const actualFiles = await this._getFilesDeep(locationFolder);
-        if (actualFiles.length > 0) {
-          return 'FILES_MISSING';
+      // Early bailout if we expect more files than what we have
+      if (expected.files.length > actual.files.length) {
+        return 'FILES_MISSING';
+      }
+
+      // Since we know the "files" array is sorted (alphabetically), we can optimize the thing
+      // Instead of storing the files in a Set, we can just iterate both arrays at once. O(n)!
+      for (let u = 0, v = 0; u < expected.files.length; ++u) {
```

Comment:
>It will always iterate on all of actual (minus advance).

Not really. It keeps incrementing `advance` as long as there are non-matching so if the non-matching files are at the start of the array, as soon as we exhaust them (`maxDiff` iterations) we stop.